src/arch/arm/64/kernel/vspace.c

```diff
@@ -647,290 +624,80 @@ exception_t checkValidIPCBuffer(vptr_t vptr, cap_t cap)
     return EXCEPTION_NONE;
 }
 
-static lookupPGDSlot_ret_t lookupPGDSlot(vspace_root_t *vspace, vptr_t vptr)
-{
-    lookupPGDSlot_ret_t ret;
-
-    pgde_t *pgd = PGDE_PTR(vspace);
-    word_t pgdIndex = GET_PGD_INDEX(vptr);
-    ret.status = EXCEPTION_NONE;
-    ret.pgdSlot = pgd + pgdIndex;
-    return ret;
-}
-
-static lookupPUDSlot_ret_t lookupPUDSlot(vspace_root_t *vspace, vptr_t vptr)
-{
-    lookupPUDSlot_ret_t ret;
-
-#ifdef AARCH64_VSPACE_S2_START_L1
-    pude_t *pud = PUDE_PTR(vspace);
-    word_t pudIndex = GET_UPUD_INDEX(vptr);
-    ret.status = EXCEPTION_NONE;
-    ret.pudSlot = pud + pudIndex;
-    return ret;
-#else
-    lookupPGDSlot_ret_t pgdSlot = lookupPGDSlot(vspace, vptr);
-
-    if (!pgde_pgde_pud_ptr_get_present(pgdSlot.pgdSlot)) {
-        current_lookup_fault = lookup_fault_missing_capability_new(PGD_INDEX_OFFSET);
-
-        ret.pudSlot = NULL;
-        ret.status = EXCEPTION_LOOKUP_FAULT;
-        return ret;
-    } else {
-        pude_t *pud;
-        pude_t *pudSlot;
-        word_t pudIndex = GET_UPUD_INDEX(vptr);
-        pud = paddr_to_pptr(pgde_pgde_pud_ptr_get_pud_base_address(pgdSlot.pgdSlot));
-        pudSlot = pud + pudIndex;
-
-        ret.status = EXCEPTION_NONE;
-        ret.pudSlot = pudSlot;
-        return ret;
-    }
-#endif
-}
-
-static lookupPDSlot_ret_t lookupPDSlot(vspace_root_t *vspace, vptr_t vptr)
-{
-    lookupPUDSlot_ret_t pudSlot;
-    lookupPDSlot_ret_t ret;
-
-    pudSlot = lookupPUDSlot(vspace, vptr);
-    if (pudSlot.status != EXCEPTION_NONE) {
-        ret.pdSlot = NULL;
-        ret.status = pudSlot.status;
-        return ret;
-    }
-    if (!pude_pude_pd_ptr_get_present(pudSlot.pudSlot)) {
-        current_lookup_fault = lookup_fault_missing_capability_new(PUD_INDEX_OFFSET);
-
-        ret.pdSlot = NULL;
-        ret.status = EXCEPTION_LOOKUP_FAULT;
-        return ret;
-    } else {
-        pde_t *pd;
-        pde_t *pdSlot;
-        word_t pdIndex = GET_PD_INDEX(vptr);
-        pd = paddr_to_pptr(pude_pude_pd_ptr_get_pd_base_address(pudSlot.pudSlot));
-        pdSlot = pd + pdIndex;
-
-        ret.status = EXCEPTION_NONE;
-        ret.pdSlot = pdSlot;
-        return ret;
-    }
-}
-
 static lookupPTSlot_ret_t lookupPTSlot(vspace_root_t *vspace, vptr_t vptr)
```

Comment:
How do we represent a lookup error in general? When

Reply:
The old lookup functions try to return a reference to the slot which can be used to inspect the state of the slot (empty, page table entry or page entry), however if the slot doesn't exist in the provided vspace because it's container pagetable doesn't exist, the reference returned is null.  A status variable is also returned to indicate whether the reference is valid or not.

The new lookup function lookupPTSlot returns the most resolved slot reference and the number of remaining bits that weren't resolved.  This way a slot reference will always be returned (because it's required to call the function with a valid vspace_root_t and a vptr_t that's contained within the vspace), and the caller is able to use the number of resolved bits return value to know how deep the slot is in the table hierarchy.